### PR TITLE
pass "nopass" for class procedure having nopass and handle optional arguments for ClassProcedure with/without nopass

### PR DESCRIPTION
**Pull request by gxyd**
pass "nopass" presence along to be used by CallBody

I've written the (only) test case with just `--show-asr`, as the `--show-llvm` will still raise an error because of a problem in `transform_optional_argument_function` pass (for details, see: https://github.com/lfortran/lfortran/issues/3632)

### CODE REVIEW

**Comment by Pranavchiku** [comment]
Can you try updating test references and see if CI passes.

File: integration_tests/module_function_with_nopass.f90 [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        if (present(x)) then
+            square_result = x * x
+        else
+            square_result = 1
+        end if
+    end function SQUARE
+
+end module calculator_mod
+
+program CalculateSquare
```
Comment (Pranavchiku) [inline integration_tests/module_function_with_nopass.f90:27]:
```suggestion
program module_function_with_nopass
```

File: integration_tests/module_function_with_nopass.f90 [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    use calculator_mod
+    implicit none
+    type(calculator) :: calc
+    integer :: number, result
+
+    number = 4
+
+    result = calc%SQUARE(number)
+    print *, 'The square of ', number, ' is ', result
+end program CalculateSquare
```
Comment (Pranavchiku) [inline integration_tests/module_function_with_nopass.f90:37]:
```suggestion
end program module_function_with_nopass
```

File: integration_tests/module_function_with_nopass.f90 [hunk 3]
```diff
@@ -0,0 +1,37 @@
+module calculator_mod
```
Comment (Pranavchiku) [inline integration_tests/module_function_with_nopass.f90:1]:
```suggestion
module modules_module_function_with_nopass
```

File: integration_tests/module_function_with_nopass.f90 [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+program CalculateSquare
+    use calculator_mod
+    implicit none
+    type(calculator) :: calc
+    integer :: number, result
+
+    number = 4
+
+    result = calc%SQUARE(number)
+    print *, 'The square of ', number, ' is ', result
```
Comment (Pranavchiku) [inline integration_tests/module_function_with_nopass.f90:36]: Let's test it with `if(abs(...)) error stop`

**Review by Pranavchiku — DISMISSED** [review]
Looks good, let's get CI working, dropped comments to improve test.

**Comment by gxyd** [comment]
> Can you try updating test references and see if CI passes.

I think it was a failure similar to the one mentioned in this issue: https://github.com/lfortran/lfortran/issues/3707. I've pushed a commit, hopefully the tests pass now.

**Comment by gxyd** [comment]
If you do like the changes, and accept it, I would request for it to be a squash the commits when merging.

**Comment by gxyd** [comment]
> I am just not sure of these nopass stuff, so I'll prefer to wait for one more approval before merging :)


Sounds good.

File: src/libasr/asr_utils.h [hunk 5]
```diff
@@ -5198,9 +5198,9 @@ static inline void Call_t_body(Allocator& al, ASR::symbol_t* a_name,
 static inline ASR::asr_t* make_FunctionCall_t_util(
     Allocator &al, const Location &a_loc, ASR::symbol_t* a_name,
     ASR::symbol_t* a_original_name, ASR::call_arg_t* a_args, size_t n_args,
-    ASR::ttype_t* a_type, ASR::expr_t* a_value, ASR::expr_t* a_dt) {
+    ASR::ttype_t* a_type, ASR::expr_t* a_value, ASR::expr_t* a_dt, bool nopass) {
 
-    Call_t_body(al, a_name, a_args, n_args, a_dt, nullptr, false, false);
+    Call_t_body(al, a_name, a_args, n_args, a_dt, nullptr, false, nopass);
```
Comment (certik) [inline src/libasr/asr_utils.h:5203]: So this is the key change --- allowing to pass through the `nopass` argument?

I think that's fine.

Comment (gxyd) [reply src/libasr/asr_utils.h:5203]: Yes

File: integration_tests/CMakeLists.txt [hunk 6]
```diff
@@ -1476,3 +1476,4 @@ RUN(NAME types_real_to_complex_cast LABELS gfortran llvm)
 RUN(NAME types_real_array_to_complex_array_cast LABELS gfortran llvm)
 
 RUN(NAME lbound_01 LABELS gfortran llvm)
+RUN(NAME module_function_with_nopass gfortran)
```
Comment (gxyd) [inline integration_tests/CMakeLists.txt:1479]: It's now tested with GFortran.

File: integration_tests/module_function_with_nopass.f90 [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    use modules_module_function_with_nopass
+    implicit none
+    type(calculator) :: calc
+    integer :: number, result
+
+    number = 4
+
+    result = calc%SQUARE(number)
+    if (abs(result) /= 1) error stop
+end program module_function_with_nopass
```
[the remark this answers is not in the document]
Comment (gxyd) [reply integration_tests/module_function_with_nopass.f90:37]: After the latest changes, it still wouldn't compile with GFortran, there is already an issue for for "optional arguments pass" here: https://github.com/lfortran/lfortran/issues/3632. I'm already working on it.

**Comment by gxyd** [comment]
@certik , this is now ready for a review. I've made more changes after your last review. We should now be able to compile the test case completely (not just till ASR), and it's tested as well (I've used `llvm2` label for it).

The issue #3632 still remains relevant (when `nopass` isn't passed to the class procedure), which depends on the changes introduced in this MR. Let me know what you think.

**Comment by gxyd** [comment]
This PR now also fixed #3632

**Comment by gxyd** [comment]
@certik , could you help me understand why the two CI checks are failing?

**Comment by Pranavchiku** [comment]
CI failure:

```console
[ 19%] Building Fortran object CMakeFiles/print_03_FAST.dir/print_03.f90.o
[ 23%] Building Fortran object CMakeFiles/print_05_FAST.dir/print_05.f90.o
make[2]: *** [CMakeFiles/module_function_with_nopass_FAST.dir/module_function_with_nopass.f90.o] Segmentation fault: 11
[ 26%] Linking Fortran executable error_stop_02_FAST
make[1]: *** [CMakeFiles/module_function_with_nopass_FAST.dir/all] Error 2
[ 30%] Linking Fortran executable error_stop_01_FAST
make[1]: *** Waiting for unfinished jobs....
[ 34%] Linking Fortran executable print_01_FAST
[ 38%] Linking Fortran executable stop_01_FAST
[ 42%] Linking Fortran executable print_03_FAST
make: *** [all] Error 2
```

**Comment by gxyd** [comment]
Ah, I see. I did see the error ofcourse, but I didn't know that you can simply run that locally by `lfortran integration_tests/module_function_with_nopass.f90 --fast`, i.e. by adding `--fast` in the command. I don't much about `--fast`, I only see: (Best performance (disable strict standard compliance), which doesn't make it much clear for me honestly.

File: integration_tests/CMakeLists.txt [hunk 8]
```diff
@@ -1478,3 +1478,5 @@ RUN(NAME types_real_to_complex_cast LABELS gfortran llvm)
 RUN(NAME types_real_array_to_complex_array_cast LABELS gfortran llvm)
 
 RUN(NAME lbound_01 LABELS gfortran llvm)
+RUN(NAME module_function_with_nopass LABELS gfortran llvm2)
+RUN(NAME module_function_without_nopass LABELS gfortran llvm2)
```
Comment (certik) [inline integration_tests/CMakeLists.txt:1482]:
```suggestion
RUN(NAME module_function_with_nopass LABELS gfortran llvm)
RUN(NAME module_function_without_nopass LABELS gfortran llvm)
```

**Comment by certik** [comment]
Related https://github.com/lfortran/lfortran/issues/3733.

**Comment by certik** [comment]
The CI failure seems unrelated: https://github.com/lfortran/lfortran/issues/3736.

**Comment by gxyd** [comment]
Yes, something along the lines of https://github.com/lfortran/lfortran/issues/3619

**Comment by gxyd** [comment]
As otherwise, they would introduce un-necessary reference files in our commit history.

**Comment by certik** [comment]
Tests pass on a rerun. I rebased and auto-squash-merged.

**Pull request merged**